.github/workflows/build-gem.yml

```diff
@@ -0,0 +1,160 @@
+name: Build gem
+
+on:
+  workflow_dispatch:
+    inputs:
+      push:
+        description: Push gem
+        required: true
+        type: boolean
+        default: true
+  push:
+    branches:
+      - "**"
+
+env:
+  GEM_HOST: 'https://rubygems.pkg.github.com/DataDog'
+
+jobs:
+  build:
+    strategy:
+      fail-fast: false
+      matrix:
+        type:
+          - final
+          - alpha
+    runs-on: ubuntu-latest
+    name: Build gem (${{ matrix.type }})
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - uses: ruby/setup-ruby@31a7f6d628878b80bc63375a93ae079ec50a1601 # v1.143.0
+        with:
+          ruby-version: '3.2'
+          bundler-cache: true # runs 'bundle install' and caches installed gems automatically
+      - name: Patch version
+        if: ${{ matrix.type != 'final' }}
+        run: |
+          # Obtain context information
+          git_ref='${{ github.ref }}'
+          git_branch="$(echo "${git_ref}" | sed -e 's#^refs/heads/##')"
+          git_sha='${{ github.sha }}'
+          gha_run_id='${{ github.run_id }}'
+
+          # Output info for CI debug
+          echo git_ref="${git_ref}"
+          echo git_branch="${git_branch}"
+          echo git_sha="${git_sha}"
+          echo gha_run_id="${gha_run_id}"
+
+          # Sanitize for ruby version usage
+          git_branch_sanitized="$(echo "$git_branch" | sed -e 's/[^a-zA-Z0-9+]\{1,\}/./g')"
+          echo git_branch_sanitized="${git_branch_sanitized}"
+
+          # Set component values:
+          # - PRE is `alpha` to be ordered before `beta` and `rc`
+          # - BUILD starts with CI run id for ordering.
+          # - BUILD has CI run id for traceability, prefixed by `gha`
+          #   for identification.
+          # - BUILD has commit next for traceability, prefixed git-describe
+          #   style by `g` for identification.
+          # - BUILD has branch name last since it has to be separated
+          #   by dots and thus has variable version segment size and
+          #   unpredictable ordering; it can thus be reliably extracted
+          #   and does not impair readability in lists
+          PRE='${{ matrix.type }}'
+          BUILD="gha${gha_run_id}.g${git_sha}.${git_branch_sanitized}"
+
+          # Output info for CI debug
+          echo PRE="${PRE}"
+          echo BUILD="${BUILD}"
+
+          # Patch in components
+          sed lib/ddtrace/version.rb -i -e "s/^\([\t ]*PRE\) *= */\1 = \'${PRE}\' # /"
+          sed lib/ddtrace/version.rb -i -e "s/^\([\t ]*BUILD\) *= */\1 = \'${BUILD}\' # /"
```

Comment:
What do you think about using ENV variables rather than manipulating files? 

```ruby
module DDTrace
  module VERSION
    MAJOR = 1
    MINOR = 11
    PATCH = 0
    PRE = ENV['PRE_VERSION'] || 'beta1'
    BUILD = ENV['BUILD_VERSION'] 

    STRING = [MAJOR, MINOR, PATCH, PRE, BUILD].compact.join('.')
end
```

The `Patch version` job could store those values, and on the `Build gem` job, we could use them. 

`PRE_VERSION=${}, BUILD_VERSION=${} bundle exec rake build`

This pattern could be used as well for `Patch gem host`

Reply:
Classic meta-programing oopsie heh :)